spdata/reader.py

```diff
@@ -1,36 +1,69 @@
-"""Various methods for reading data"""
+"""Various methods for reading data
 
-from typing import List, Tuple
-import numpy as np
+Copyright 2018 Spectre Team
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import os
 
-import spdata.types as ty
+from typing import List, Tuple, Union
+
+import numpy as np
 import pyimzml.ImzMLParser as imzparse
 
+from . import types as ty
+from . import discover as disc
+from .common import Name, Path
+
 def _parse_metadata(line: str) -> (int, int, int, int):
     x, y, z, label, *_ = line.split()
     return int(x), int(y), int(z), int(label)
 
-
 def _parse_data(line: str) -> List[float]:
     return [float(value) for value in line.split()]
 
-
 def _load_entry(metadata_line: str, data_line: str) -> Tuple[
         Tuple[int, int, int], int, List[float]]:
     x, y, z, label = _parse_metadata(metadata_line)
     data = _parse_data(data_line)
     return (x, y, z), label, data
 
-
-def load_txt(content) -> ty.Dataset:
+# Definition of loaders
+from functools import wraps
+loaders = {}
+
+def Loader(ext: str):
+    def register_loader(f):
+        loaders.setdefault(ext, f)
+        @wraps(f)
+        def loader_wrapper(file_path: Path):
+            return f(file_path)
+        return loader_wrapper
+    return register_loader
+
+@Loader('.txt')
+def load_txt(file_path: Path) -> ty.Dataset:
     """Load Dataset from file
 
     Args:
-        content (iterable): data file content
+        file_path : data file path
 
     Returns:
         out: spdata.types.Dataset
     """
+    content = os.open(file_path)
+
```

Comment:
This leaves the file open; please use the proper idiom utilizing a context manager:
```py
with open(path) as f:
    content = f.read()   # or read_lines()
```
Alternatively, there should be a `try ... finally ...` construct, with guaranteed `content.close()` (not recommended).